Bluetooth: controller: Add/Remove ACAD on create/terminate BIG

cvinayak opened the pull request:
Added implementation to add/remove ACAD field in the common
extended header format of the periodic advertising PDU on
create/terminate BIG.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add/Remove ACAD on create/terminate BIG

Added implementation to add/remove ACAD field in the common
extended header format of the periodic advertising PDU on
create/terminate BIG.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: ISO test with ACAD in periodic advertising

Updated ISO test to demonstrate ACAD field in periodic
advertising PDUs. Here, test changing the periodic
advertising data.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: name adv_handle for advertising handle

Use the name adv_handle instead of index to store advertising
handle.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Use set/clear to add/remove ACAD

Use the set/clear function to modify the common extended
header format in the PDU to add/remove ACAD.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -763,6 +769,21 @@ static uint8_t adv_sync_hdr_set_clear(struct lll_adv_sync *lll_sync,
+		acad_len = *(uint8_t *)value;
+		value = (uint8_t *)value + 1;
+		/* return the pointer to ACAD offset */
+		sys_put_le32((uint32_t)ter_dptr, value);
+		value = (uint8_t *)value + sizeof(void *);
```
` sizeof(value)` ?

wopu-ot replied:
Incrementing by `sizeof(uint32)` would match what is stored above and would work if `sizeof(void *)` <= `sizeof(uint32)`. If `sizeof(void *) > `sizeof(uint32)`, storing the pointer as `uint32_t` is broken in the first place.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -130,6 +135,22 @@ uint8_t ll_big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bis,
 		return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
 	}
+
+	/* Add ACAD to AUX_SYNC_IND */
+	hdr_data.field_data = field_data;
+	field_data[0] = sizeof(struct pdu_big_info) + 2;
+	err = ull_adv_sync_pdu_set_clear(adv, ULL_ADV_PDU_HDR_FIELD_ACAD, 0U,
+					 &hdr_data, &ter_idx);
+	if (err) {
+		return err;
+	}
+
+	acad = (void *)sys_get_le32(&field_data[1]);
+	acad[0] = sizeof(struct pdu_big_info) + 1;
+	acad[1] = BT_DATA_BIG_INFO;
```
This is where it is going in terms of implementation: https://github.com/zephyrproject-rtos/zephyr/pull/31009/commits/02e62ba4f365e29739006500f28c788436a0ade1#diff-46edd4480ca72c05f7bb026bdac384a2d48b03317361f2ea6ccdf70d15dc6769R197-R241

Passing by parameters or filling into a structure or passing the whole LLL context to fill inside `set_clear`, all have their disadvantage.

wopu-ot replied:
I think having dedicated parameters or fields for passing the length and returning the address instead of reusing `field_data` would clear up a lot. Maybe the general pattern should be that `field_data` does not hold the data to be filled in on the way down, but the address where to write data on the way up.

So instead of writing (just as an example):
```
                cte_info.type = df_cfg->cte_type;
                cte_info.time = df_cfg->cte_length;
                pdu_data.field_data = (uint8_t *)&cte_info;
                ...
                err = ull_adv_sync_pdu_set_clear(adv, ULL_ADV_PDU_HDR_FIELD_CTE_INFO, 0, &pdu_data, &ter_idx);
```
the pattern could be
```
                err = ull_adv_sync_pdu_set_clear(adv, ULL_ADV_PDU_HDR_FIELD_CTE_INFO, 0, &pdu_data, &ter_idx);
                cte_info = pdu_data.field_data;                
                cte_info->type = df_cfg->cte_type;
                cte_info->time = df_cfg->cte_length;
```
This would then also fit the use case for the somewhat bigger big_info struct.

cvinayak replied:
The intention is to use the `ull_adv_aux_hdr_set_clear` and `ull_adv_sync_pdu_set_clear` (in the future combine them into one function) to take-in variable number of data for numerous header fields (bitmap for set/clear, say adding adv and init addresses, may be other parameters together). Hence the field_data is length encoded (like the AD data format, length-data). This is done to save on execution time updating the PDU.

Let me review so far, if array of lengths can be passed and then return array of pointers that can be used to fill in. (But this will not happen until I get to the point I have an initial samples of ISO broadcaster and ISO receiver)

The current implementation of passing length+value for simple values (AdvA, TargetA, ADI, AuxPtr, TxPower and Host AD Data) and returning pointers for struct values (CTEInfo, SyncInfo, and ACAD).

This means, we keep as-is for AdvA, TargetA, ADI, AuxPtr, TxPower and Host AD Data; and CTEInfo and SyncInfo be ported to populating the return address, do you agree?

wopu-ot replied:
I would prefer to use the same pattern for everything and I think it would eventually lead to simpler, more efficient code. But I can live with distinguishing between "simple" and "complex" cases.

cvinayak replied:
@wopu-ot as communicated to you few weeks back in our status meeting, I tend to agree with passing down field bitmap and passing up address as stated in your comment here https://github.com/zephyrproject-rtos/zephyr/pull/32622#discussion_r587425119

@andrzej-kaczmarek FYI.

cvinayak replied:
@wopu-ot Updated with the sync_info population by caller. Please review.

cvinayak committed:
Bluetooth: controller: Refactor sync_info population

Based on review comments, refactor out sync_info population
to be performed by the caller of the function that prepares
the extended advertising PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -163,6 +163,10 @@ void ull_adv_sync_update(struct ll_adv_sync_set *sync, uint32_t slot_plus_us,
 
+/* helper function to reserve ACAD field in PDU buffer */
+uint8_t ull_adv_sync_acad_enable(struct lll_adv_sync *lll_sync,
+				 uint8_t acad_len, void **acad);
```
I'm curious as to why the `acad` field here is a `void **` rather than `uint8_t **`. Can you elaborate?

cvinayak replied:
Isnt this outdated code.

`void *` is guaranteed by compilers to be assigned to pointers of other data types without explicit typecast (similar to why `malloc` returns `void *`).

Thalley replied:
It is. I was reviewing commit-by-commit. You can disregard this.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -130,6 +132,17 @@ uint8_t ll_big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bis,
+
+	/* Add ACAD to AUX_SYNC_IND */
+	err = ull_adv_sync_acad_enable(lll_adv_sync,
+				       (sizeof(struct pdu_big_info) + 2),
```
I'm assuming that the `+ 2` is for the length field and AD type. Don't we have any AD header struct that we can call the `sizeof()` on, or a macro that better describes this value we can use instead?

cvinayak replied:
Isnt this old code?

Thalley replied:
It is. I was reviewing commit-by-commit. You can disregard this.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -130,6 +132,17 @@ uint8_t ll_big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bis,
+	if (err) {
+		return err;
+	}
+	acad[0] = sizeof(struct pdu_big_info) + 1;
+	acad[1] = BT_DATA_BIG_INFO;
```
Would it make sense to also add the AD type to the list of parameters for `ull_adv_sync_acad_enable`? It seem that if we do that, then `ull_adv_sync_acad_enable` could be responsible for settings the AD type as well as the length field.

cvinayak replied:
ACAD can have mutiple advertising data types in a length encoded format. Hence, it is best left to the caller of the function to populate the ACAD with multiple data types.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -554,6 +554,151 @@ uint8_t ull_adv_sync_pdu_set_clear(struct ll_adv_set *adv,
+
+	/* Get reference to previous tertiary PDU data */
+	ter_pdu_prev = lll_adv_sync_data_peek(lll_sync, NULL);
+	ter_com_hdr_prev = (void *)&ter_pdu_prev->adv_ext_ind;
```
I feel like I've asked this question before, but what was the reason to cast to a `(void *)` instead of `(struct pdu_adv_com_ext_adv *)`?

cvinayak replied:
`void *` being and is guaranteed by compilers to be assigned to pointers of other data types without explicit typecast (similar to why `malloc` returns `void *`).

Thalley replied:
> void * being and is guaranteed by compilers to be assigned to pointers of other data types

Do you mean assignable instead of assigned? If so, that makes sense.

cvinayak replied:
yes, assignable.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -554,6 +554,151 @@ uint8_t ull_adv_sync_pdu_set_clear(struct ll_adv_set *adv,
+	ter_com_hdr->adv_mode = ter_com_hdr_prev->adv_mode;
+	ter_hdr = (void *)ter_com_hdr->ext_hdr_adv_data;
+	ter_dptr = ter_hdr->data;
+	*(uint8_t *)ter_hdr = 0U;
```
What is the reasoning behind only setting the first octet to 0, instead of e.g. doing a memset?

cvinayak replied:
If extended header is present, the minimum 1 octet is the flags.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -50,13 +50,14 @@ const uint8_t *ull_adv_pdu_update_addrs(struct ll_adv_set *adv,
  */
-struct adv_pdu_field_data {
-	uint8_t *field_data;
+struct ull_adv_ext_hdr_data {
+	void *field_data;
```
Can you elaborate on why this type was changed?

cvinayak replied:
to be consistent with `extra_data` below.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -134,15 +137,20 @@ uint8_t ll_big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bis,
-				       (sizeof(struct pdu_big_info) + 2),
-				       (void **)&acad);
+	hdr_data.field_data = field_data;
+	field_data[0] = sizeof(struct pdu_big_info) + 2;
```
Similar comment ragarding the size of the advertising data header

cvinayak replied:
No, there is no define and host too uses magic value 2 (https://github.com/zephyrproject-rtos/zephyr/blob/aec4c4977dbccef2d4486eb37d12ea679cb9d9c8/subsys/bluetooth/host/adv.c#L315)

As a separate task, may be develop some reusable code for AD data population.

Thalley replied:
Since we have no precedence of using a macro value for this, I'll let it pass, but it's nonetheless something we generally should avoid doing in my opinion.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -611,16 +607,30 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
+		 * buffer
+		 */
+		sys_put_le32((uint32_t)sec_dptr, value);
+		value = (uint8_t *)value + sizeof(void *);
```
I guess the same comment regarding the size of `void *` also goes here

cvinayak replied:
changed to `uint32_t` which is being filed into `value` pointer.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -433,6 +443,34 @@ void ull_adv_sync_release(struct ll_adv_sync_set *sync)
+	lll_sync = &sync->lll;
+	memcpy(si->sca_chm, lll_sync->data_chan_map,
+	       sizeof(si->sca_chm));
+	si->sca_chm[4] &= 0x1f;
```
I assume that `0x1F` is some kind of mask that is applied. We don't have a `#define` for that anywhere?

cvinayak replied:
SCA is encoded as 3 MS bits in the sync_info structure's channel map array. 0x1F is used as part of set operation.

Thalley replied:
Hmm, let me see if I understand this correctly. 

The is a 5 octec (40 bit) field containing the Channel Map (ChM) and the SCA. We apply the 0x1F mask on the "highest" octet so that we set the SCA value to 0 by doing `=& 0x1F`, and then by
```
si->sca_chm[4] |= lll_clock_sca_local_get() << 5;
```
We set the 3 bit SCA value to what `lll_clock_sca_local_get()` returns. 

Is the above correct? 

Defining this mask isn't strictly necessary, but I general prefer it.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_iso/src/main.c:
```diff
@@ -85,13 +96,15 @@ static void test_iso_main(void)
+	printk("success.\n");
 
-	/* Start extended advertising */
+	printk("Start extended advertising...\n");
```
Typically you don't have the `\n` when you do these `Doing x... success` printk's

cvinayak replied:
Fixed.

Thalley reviewed (approved):
There are some possible readability additions that could be applied, but since there's not really any precedence set for the controller for the specified lines I don't think it's strictly necessary to change, but something to keep in mind.

cvinayak committed:
Bluetooth: Controller: Minor fix debug message with channel map mask

Minor fix to channel map mask used in debug message to print
sync info fields.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use macros for SCA and Channel Map access

Use macros to access SCA and Channel Map fields in the Sync
Info structure in advertising PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@wopu-ot Could you take another look.

wopu-ot reviewed (changes requested):
The change still passes pointers through a 4-byte/uint32_t value, which may work right now but is a bug that will manifest sooner or later. Please update to use appropriately typed fields and variables.

cvinayak commented:
> 
> 
> The change still passes pointers through a 4-byte/uint32_t value, which may work right now but is a bug that will manifest sooner or later. Please update to use appropriately typed fields and variables.

I have replaced to use memcpy, please review.

cvinayak commented:
Also converted to draft as I want https://github.com/zephyrproject-rtos/zephyr/pull/31875 to be merged before this, to ensure I can rebase this PR thereafter

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -768,6 +813,21 @@ static uint8_t adv_sync_hdr_set_clear(struct lll_adv_sync *lll_sync,
+		acad_len = *(uint8_t *)value;
+		value = (uint8_t *)value + 1;
+		/* return the pointer to ACAD offset */
+		memcpy(value, &ter_dptr, sizeof(ter_dptr));
+		value = (uint8_t *)value + sizeof(ter_dptr);
```
@wopu-ot This should cover pointer size, right?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1208,14 +1209,24 @@ uint8_t ll_adv_enable(uint8_t enable)
+				 * followed by pointer to sync_info in the
+				 * PDU.
+				 */
+				sync_info = (void *)sys_get_le32(&value[1]);
```
will fix this and send updates

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -263,23 +264,32 @@ uint8_t ll_adv_sync_enable(uint8_t handle, uint8_t enable)
+		 * sync_info structure, followed by pointer to sync_info in the
+		 * PDU.
+		 */
+		sync_info = (void *)sys_get_le32(&value[1]);
```
will fix this and send updates
